### PR TITLE
aws: Supply cluster manager ref to aws credentials provider

### PR DESCRIPTION
**Pull request by suniltheta**
Signed-off-by: Sunil Narasimhamurthy <sunnrs@amazon.com>

Commit Message: aws: Supply cluster manager ref to aws credentials provider
Additional Description: Inside the aws credentials provider we are going to create a static DNS cluster. This way the users don't have to explicitly supply the cluster configuration, similar to how user won't do today. 
In this PR we are just sending Cluster Manager context inside the AWS Credentials provider so that we can explicitly call `addOrUpdateCluster` which will be introduced in the subsequent CR.

Risk Level: Low
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/common/grpc/google_grpc_utils.cc [hunk 1]
```diff
@@ -133,8 +133,10 @@ GoogleGrpcUtils::channelArgsFromConfig(const envoy::config::core::v3::GrpcServic
 }
 
 std::shared_ptr<grpc::Channel>
-GoogleGrpcUtils::createChannel(const envoy::config::core::v3::GrpcService& config, Api::Api& api) {
-  std::shared_ptr<grpc::ChannelCredentials> creds = getGoogleGrpcChannelCredentials(config, api);
+GoogleGrpcUtils::createChannel(const envoy::config::core::v3::GrpcService& config, Api::Api& api,
```
Comment (abaptiste) [inline source/common/grpc/google_grpc_utils.cc:136]: I think you might get push back here.  If a customer is using the grpc channel for xDS and *not* using AWS auth, they still need to pass in a cluster manager. 

Is this the only way to get the cluster manager into the signer?

Comment (suniltheta) [reply source/common/grpc/google_grpc_utils.cc:136]: Yes, they still need to pass in a cluster manager even if they are not using AWS auth :( 

In `source/common/grpc/async_client_manager_impl.cc` L123 func `AsyncClientManagerImpl::factoryForGrpcService`, the switch case `kEnvoyGrpc (AsyncClientFactoryImpl)` was taking in the cluster manager `cm_` but not `kGoogleGrpc (GoogleAsyncClientFactoryImpl)`. With this change if I need the cluster manager I am forced to send the cluster manager all the way from `GoogleAsyncClientFactoryImpl` into AWS CredentialProviders.

Comment (suniltheta) [reply source/common/grpc/google_grpc_utils.cc:136]: as of now I haven't discovered an alternate.

**Review by abaptiste — CHANGES_REQUESTED** [review]
Is there an alternate way of getting the cluster manager from a "global" or context object within the filter?

**Comment by suniltheta** [comment]
I would also prefer that option of getting from a global, let me see if some one has an answer on Envoy slack

File: source/common/grpc/google_grpc_creds_impl.cc [hunk 2]
```diff
@@ -140,7 +140,7 @@ class DefaultGoogleGrpcCredentialsFactory : public GoogleGrpcCredentialsFactory
 public:
   std::shared_ptr<grpc::ChannelCredentials>
   getChannelCredentials(const envoy::config::core::v3::GrpcService& grpc_service_config,
-                        Api::Api& api) override {
+                        Api::Api& api, Upstream::ClusterManager&) override {
```
Comment (jmarantz) [inline source/common/grpc/google_grpc_creds_impl.cc:143]: drive-by comment; is the end goal of this giant plumbing exercise just to get the cluster-manager here?

is that something that will eventually be used?

Comment (suniltheta) [reply source/common/grpc/google_grpc_creds_impl.cc:143]: Yes all these effort is to get the cluster manager inside source/extensions/grpc_credentials/aws_iam/config.cc

It will eventually be used in my follow up CRs to create internal cluster 

For example like here: https://github.com/suniltheta/envoy/blob/cm_ref_backup/source/extensions/common/aws/utility.cc#L325

Comment (jmarantz) [reply source/common/grpc/google_grpc_creds_impl.cc:143]: For the most part I think the approach you want to take here is not too far from what you've done. But I suspect there's a better path that doesn't involve using global or static variables.

If your config functions in aws_iam/config.cc can be made part of a class, and that can be constructed at configuration time, then I think it'll be easier for you to plumb the CM in that way, rather than via grpc.

I don't totally understand why the CM is needed for you to bypass curl, but if you did need the CM I think you could get it when you construct your aws utility object.

Right now you might not *have* an AWS utility object (I haven't checked) but I suspect that would make things easier. Or maybe there's another context you can latch onto that you control.

Comment (suniltheta) [reply source/common/grpc/google_grpc_creds_impl.cc:143]: As of today for aws utility we are not creating any objects. We are just utilizing static functions inside `aws::Utility` (ref: [utility.h](https://github.com/envoyproxy/envoy/blob/main/source/extensions/common/aws/utility.h)).
I am not able to readily find a `context` that I can utilize inside aws utility. This approach would have been much simpler.

Do you happen to know an example where cm is obtained from a context?

---

I don't have much knowledge on the core components of envoy code, but would making use of `Registry::FactoryRegistry` acceptable? I don't have a solution in mind, I just want to make sure if I can proceed in this direction.

> why the CM is needed

As of now I am trying to model aws metadata fetcher similar to [jwks_fetcher.cc](https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/common/jwks_fetcher.cc#L40). The CM will be used to get the threadLocalCluster to make httpAsync calls.
